pkg/pipeline/pipeline_builder.go

```diff
@@ -0,0 +1,335 @@
+			err = fmt.Errorf("invalid stage type: %v", pEntry.stageType)
+		}
+		if err != nil {
+			return err
+		}
+		b.pipelineEntryMap[stage.Name] = &pEntry
+		b.pipelineStages = append(b.pipelineStages, &pEntry)
+		log.Debugf("pipeline = %v", b.pipelineStages)
+	}
+	log.Debugf("pipeline = %v", b.pipelineStages)
+	return nil
+}
+
+// reads the configured Go stages and connects between them
+func (b *builder) build() (*Pipeline, error) {
+	if err := b.readStages(); err != nil {
+		return nil, err
+	}
+
+	for _, connection := range b.configStages {
+		if connection.Name == "" || connection.Follows == "" {
+			// ignore entries that do not represent a connection
+			continue
+		}
+		// instantiates (or loads from cache) the destination node of a connection
+		dstEntry, ok := b.pipelineEntryMap[connection.Name]
+		if !ok {
+			return nil, fmt.Errorf("unknown pipeline stage: %s", connection.Name)
+		}
+		dstNode, err := b.getStageNode(dstEntry)
+		if err != nil {
+			return nil, err
+		}
+		dst, ok := dstNode.(node.Receiver)
+		if !ok {
+			return nil, fmt.Errorf("stage %q of type %q can't receive data",
+				connection.Name, dstEntry.stageType)
+		}
+		// instantiates (or loads from cache) the source node of a connection
+		srcEntry, ok := b.pipelineEntryMap[connection.Follows]
+		if !ok {
+			return nil, fmt.Errorf("unknown pipeline stage: %s", connection.Follows)
+		}
+		srcNode, err := b.getStageNode(srcEntry)
+		if err != nil {
+			return nil, err
+		}
+		src, ok := srcNode.(node.Sender)
+		if !ok {
+			return nil, fmt.Errorf("stage %q of type %q can't send data",
+				connection.Name, dstEntry.stageType)
+		}
+		log.Debugf("connecting stages: %s --> %s", connection.Follows, connection.Name)
+
+		// connects source and destination node, and catches any panic from the Go-Pipes library.
+		var catchErr error
+		func() {
+			defer func() {
+				if msg := recover(); msg != nil {
+					catchErr = fmt.Errorf("%q and %q stages haven't compatible input/outputs: %v",
+						connection.Follows, connection.Name, msg)
+				}
+			}()
+			src.SendsTo(dst)
+		}()
+		if catchErr != nil {
+			return nil, catchErr
+		}
+	}
+
+	if len(b.startNodes) == 0 {
+		return nil, errors.New("no ingesters have been defined")
+	}
+	if len(b.terminalNodes) == 0 {
+		return nil, errors.New("no writers have been defined")
+	}
+	return &Pipeline{
+		startNodes:     b.startNodes,
+		terminalNodes:  b.terminalNodes,
+		pipelineStages: b.pipelineStages,
+	}, nil
+}
+
+func (b *builder) getStageNode(pe *pipelineEntry) (interface{}, error) {
+	if stg, ok := b.createdStages[pe.configStage.Name]; ok {
+		return stg, nil
+	}
+	var stage interface{}
+	// TODO: modify all the types' interfaces to not need to write loops here, the same
+	// as we do with Ingest
+	switch pe.stageType {
+	case StageIngest:
+		init := node.AsInit(pe.Ingester.Ingest)
+		b.startNodes = append(b.startNodes, init)
+		stage = init
+	case StageWrite:
+		term := node.AsTerminal(func(in <-chan []config.GenericMap) {
```

Comment:
It might be convenient to allow write_stdout at any step of the pipeline, allow it multiple times, and allow it to have followers. This way we can debug and have write_stdout after each other stage (decode, transform1, transform2, etc) in a pipeline. This is related to what @eranra mentioned with the possibility of (write, export) nodes being sometimes terminal nodes and sometimes middle nodes.

Reply:
We could modify the interface so it can fit into a middle node instead of a terminal mode. I'd do it in another PR, to not loose the focus on the goal of this PR.